Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp

```diff
@@ -1378,10 +1378,47 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionStructure(GUniquePtr<Gst
     GST_TRACE("emitted decryption structure on pipeline, event handled %s", boolForPrinting(eventHandled));
 }
 
-void MediaPlayerPrivateGStreamerBase::reportWaitingForKey()
+void MediaPlayerPrivateGStreamerBase::setWaitingForKey(bool waitingForKey)
 {
-    GST_TRACE("waiting for key");
-    m_player->waitingForKey();
+    // We bail out if values did not change or if we are requested to not wait anymore but there are still waiting decryptors.
+    GST_TRACE("waitingForKey %s, m_waitingForKey %s", boolForPrinting(waitingForKey), boolForPrinting(m_waitingForKey));
+    if (waitingForKey == m_waitingForKey || (!waitingForKey && this->waitingForKey()))
+        return;
+
+    m_waitingForKey = waitingForKey;
+    GST_DEBUG("waiting for key changed %s", boolForPrinting(m_waitingForKey));
+    m_player->waitingForKeyChanged();
+}
+
+bool MediaPlayerPrivateGStreamerBase::waitingForKey() const
+{
+    bool result { false };
+    if (m_pipeline) {
+        GstState state;
+        gst_element_get_state(m_pipeline.get(), &state, nullptr, 0);
+        if (state >= GST_STATE_PAUSED) {
+            GRefPtr<GstQuery> query = adoptGRef(gst_query_new_custom(GST_QUERY_CUSTOM, gst_structure_new_empty("any-decryptor-waiting-for-key")));
+            result = gst_element_query(m_pipeline.get(), query.get());
+            GST_TRACE("query result %s, on %s", boolForPrinting(result), gst_element_state_get_name(state));
+        } else if (state >= GST_STATE_READY) {
+            // Running a query is more straightforward but it only works when the pipeline is set up and running, otherwise we need to inspect it and ask the decryptors manually.
+            GUniquePtr<GstIterator> iterator(gst_bin_iterate_recurse(GST_BIN(m_pipeline.get())));
+            GstIteratorResult iteratorResult;
+            do {
+                iteratorResult = gst_iterator_fold(iterator.get(), [](const GValue *item, GValue *, gpointer) -> gboolean {
+                    GstElement* element = GST_ELEMENT(g_value_get_object(item));
+                    return !WEBKIT_IS_MEDIA_CENC_DECRYPT(element) || !webKitMediaCommonEncryptionDecryptIsWaitingForKey(WEBKIT_MEDIA_CENC_DECRYPT(element));
```

Comment:
This would read more naturally as return IS_CENC_DECRYPT(element) && isWaitingForKey(element)

Reply:
Yes, but we need to keep the null namespacing somehow.

Reply:
I don't understand, I'm applying de morgan here, it would be nice to applying a negation across the conditional and switch it to &&

Reply:
Me no entiende what you say either :upside_down_face: .

We are doing something similar to what we do with query, we are coalescing the return value of the query being handled and the return value of the query itself because for us it is ok if we find just 1 decryptor that is still blocked to answer `true` so we if just answer `true`, it means that we can stop the query and it means we found a waiting decryptor.

For the case of the iterator, the iterator fold continues when we return `TRUE` in and we want to stop only if we find a decryptor and that decryptor is waiting, so we applying De Morgan we continue if we are not dealing with a decryptor or that decryptor is not waiting.

Reply:
I think I got it now, so you prefer to negate the whole thing and have an and? I don't :smile: